### ISSUE TITLE
Filenames containing a colon fail to sync

### ISSUE THREAD

**Issue by thomasfedb** [skilion/onedrive#95]
Any file which contains a colon in the filename fails to sync with a 400 error logged.

```
Aug 30 10:03:04 zephon onedrive[12033]: Uploading: ./Research/Publications/2016 - Wilson - Pain Profiles: Head and Neck.pdf
Aug 30 10:03:06 zephon onedrive[12033]: HTTP request returned status code 400 (Bad Request)
```

```
Sep 01 19:05:04 zephon onedrive[19249]: Uploading: ./foo: bla.txt
Sep 01 19:05:05 zephon onedrive[19249]: HTTP request returned status code 400 (Bad Request)
```

It also appears that after sever attemps to sync such a file the process with segfault.

```
Aug 30 10:02:53 zephon onedrive[12012]: sync.SyncException@src/sync.d(352): HTTP request returned status code 400 (Bad Request)
Aug 30 10:02:53 zephon onedrive[12012]: ----------------
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void sync.SyncEngine.scanForDifferences(immutable(char)[]) [0x59aa9c]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void main.performSync(sync.SyncEngine) [0x58b03e]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? _Dmain [0x58a797]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv [0x5a57ce]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x5a5718]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() [0x5a578a]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x5a5718]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? _d_run_main [0x5a5689]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? main [0x58cc95]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? __libc_start_main [0x354e06ff]
Aug 30 10:02:53 zephon onedrive[12012]: onedrive.OneDriveException@src/onedrive.d(338): HTTP request returned status code 400 (Bad Request)
Aug 30 10:02:53 zephon onedrive[12012]: ----------------
Aug 30 10:02:53 zephon onedrive[12012]: ??:? std.json.JSONValue onedrive.OneDriveApi.upload(immutable(char)[], immutable(char)[]) [0x591967]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void sync.SyncEngine.uploadNewFile(immutable(char)[]) [0x59cc35]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0x59c686]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0x59c4bf]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0x59c4bf]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void sync.SyncEngine.uploadNewItems(immutable(char)[]) [0x59c4bf]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void sync.SyncEngine.scanForDifferences(immutable(char)[]) [0x59a99b]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void main.performSync(sync.SyncEngine) [0x58b03e]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? _Dmain [0x58a797]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv [0x5a57ce]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x5a5718]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() [0x5a578a]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x5a5718]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? _d_run_main [0x5a5689]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? main [0x58cc95]
Aug 30 10:02:53 zephon onedrive[12012]: ??:? __libc_start_main [0x354e06ff]
```

**Comment by skilion** [comment]
The colon is a reserved character in Windows.
See https://msdn.microsoft.com/en-us/library/aa365247

**Comment by thomasfedb** [comment]
Hey @skilion, I have discovered this fact also, however I'm not sure that segfaulting should be considered appropriate behaviour in this condition.

Perhaps an error message?

Ideally filenames which are illegal on OneDrive would be transformed to a legal name, although I understand if that's outside of the scope of this project.

**Comment by Gr1mR3aver** [comment]
I tend to agree that segfault is not the answer.  I discovered a similar issue where the app segfaults if it can't update the mtimes/atime on a file.  I just added a workaround to ignore "FileException" failures when trying to set the mtime/atime.

In any case, I think the addition more graceful exception handling, with appropriate logging, would be better.  In this particular case, I think a failure, with a log message along the lines of "Illegal filename, file skipped" would be appropriate.  I am not much of a dev or I'd submit the patch myself.  I can still take a swing at it if @skilion is willing to take a look.
